p2p/net/upgrader/upgrader.go

```diff
@@ -74,15 +75,21 @@ type upgrader struct {
 	//
 	// If unset, the default value (15s) is used.
 	acceptTimeout time.Duration
+
+	msmuxer    *mss.MultistreamMuxer //
+	muxers     map[string]network.Multiplexer
+	preference []string
 }
 
 var _ transport.Upgrader = &upgrader{}
 
-func New(secureMuxer sec.SecureMuxer, muxer network.Multiplexer, opts ...Option) (transport.Upgrader, error) {
+func New(secureMuxer sec.SecureMuxer, muxers map[string]network.Multiplexer, preference []string, opts ...Option) (transport.Upgrader, error) {
```

Comment:
These constructor options are very unclean.

What you want to use here is list of `<id>, <muxer>`, ordered by preference. Introduce a type that reflects that.